### DISCUSSION TITLE
has this library built-in cache?

### DISCUSSION THREAD

**Discussion by Javier-Szyfer** [wevm/wagmi#40] [Q&A] [answered]
Hey guys, sorry if this is a noob question, but im wondering if theres already caching built in like in swr hooks .
I'm using useContractRead to query an array in my contract (+ formatting data) and even though  te hook response is fast , when i map and format the data it takes a while .. so maybe (if im not missing something..) it would be great to implement some kind of caching stale-while-revalidate . 
Otherwise does anyone have an idea how to cache the data after the first request? I am still struggling to make a subgraph  
Thanks so much for this.

**Comment by iduuck** [comment, marked as answer]
No, not (yet?) – we are currently only using the state for response. BUT, you could connect swr and wagmi to get the best of both worlds. Since swr is accepting promises, you could easily use the `read` function, which is being returned by the `useContractRead` and let swr cache it inside it's system. 

```ts
const ComponentName = () => {
  // example from docs: https://wagmi-xyz.vercel.app/docs/hooks/useContractRead#usage
  const [_, read] = useContractRead(
    {
      addressOrName: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
      contractInterface: wagmigotchiABI,
    },
    'getHunger',
  )

  // example from swr docs: https://swr.vercel.app/docs/data-fetching#fetch
  const { data } = useSWR('some-unique-id', read)

  // compoent here...
}
```

or you could go even further and create your own custom hook, if you need the caching on multiple positions:

```ts
const useCachedContractRead = () => {
  // example from docs: https://wagmi-xyz.vercel.app/docs/hooks/useContractRead#usage
  const [_, read] = useContractRead(
    {
      addressOrName: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
      contractInterface: wagmigotchiABI,
    },
    'getHunger',
  )

  // example from swr docs: https://swr.vercel.app/docs/data-fetching#fetch
  return useSWR('some-unique-id', read)
}
```

And then you can just use the hook like before with `useSWR`.

**Reply by Javier-Szyfer** [reply]
thanks so much for your quick reply, will definitely try this : )
